homeassistant/components/reolink/media_source.py

```diff
@@ -0,0 +1,329 @@
+        )
+
+    async def _async_generate_camera_days(
+        self, config_entry_id: str, channel: int, stream: str
+    ) -> BrowseMediaSource:
+        """Return all days on which recordings are available for a reolink camera."""
+        host = self.data[config_entry_id].host
+
+        # We want today of the camera, not necessarily today of the server
+        now = host.api.time() or await host.api.async_get_time()
+        start = now - dt.timedelta(days=31)
+        end = now
+
+        children: list[BrowseMediaSource] = []
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug(
+                "Requesting recording days of %s from %s to %s",
+                host.api.camera_name(channel),
+                start,
+                end,
+            )
+        statuses, _ = await host.api.request_vod_files(
+            channel, start, end, status_only=True, stream=stream
+        )
+        for status in statuses:
+            for day in status.days:
+                children.append(
+                    BrowseMediaSource(
+                        domain=DOMAIN,
+                        identifier=f"DAY|{config_entry_id}|{channel}|{stream}|{status.year}|{status.month}|{day}",
+                        media_class=MediaClass.DIRECTORY,
+                        media_content_type=MediaType.PLAYLIST,
+                        title=f"{status.year}/{status.month}/{day}",
+                        can_play=False,
+                        can_expand=True,
+                    )
+                )
+
+        return BrowseMediaSource(
+            domain=DOMAIN,
+            identifier=f"DAYS|{config_entry_id}|{channel}|{stream}",
+            media_class=MediaClass.CHANNEL,
+            media_content_type=MediaType.PLAYLIST,
+            title=f"{host.api.camera_name(channel)} {res_name(stream)}",
+            can_play=False,
+            can_expand=True,
+            children=children,
+        )
+
+    async def _async_generate_camera_files(
+        self,
+        config_entry_id: str,
+        channel: int,
+        stream: str,
+        year: int,
+        month: int,
+        day: int,
+    ) -> BrowseMediaSource:
+        """Return all recording files on a specific day of a Reolink camera."""
+        host = self.data[config_entry_id].host
+
+        start = dt.datetime(year, month, day, hour=0, minute=0, second=0)
+        end = dt.datetime(year, month, day, hour=23, minute=59, second=59)
+
+        children: list[BrowseMediaSource] = []
+        if _LOGGER.isEnabledFor(logging.DEBUG):
+            _LOGGER.debug(
+                "Requesting VODs of %s on %s/%s/%s",
+                host.api.camera_name(channel),
+                year,
+                month,
+                day,
+            )
+        _, vod_files = await host.api.request_vod_files(
+            channel, start, end, stream=stream
+        )
+        for file in vod_files:
+            file_name = f"{file.start_time.time()} {file.duration}"
```

Comment:
The time conversion here is likely expensive but I don't think its avoidable

Reply:
yea, I don't really think we can avoid it.
Upstream we get a time as string, but then we add timezone info to it and make it a datetime in order to correct for the timezone offsets. Here we need to go back to a string...